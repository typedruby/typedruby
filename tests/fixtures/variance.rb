# @typedruby

class Invariant::[T]
  def initialize(T v) => nil; end
end

class Covariant::[+T]
  def initialize(T v) => nil; end
end

class Contravariant::[-T]
  def initialize(T v) => nil; end
end

class A
end

class B < A
end

class C < B
end

def inv(Invariant::[B] x) => nil; end

def test_inv => nil
  inv(Invariant.new(A.new))
  inv(Invariant.new(B.new))
  inv(Invariant.new(C.new))
end

def cov(Covariant::[B] x) => nil; end

def test_cov => nil
  cov(Covariant.new(A.new))
  cov(Covariant.new(B.new))
  cov(Covariant.new(C.new))
end

def contra(Contravariant::[B] x) => nil; end

def test_contra => nil
  contra(Contravariant.new(A.new))
  contra(Contravariant.new(B.new))
  contra(Contravariant.new(C.new))
end
