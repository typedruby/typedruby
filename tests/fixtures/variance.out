error: Could not match types:

        @ __ROOT__/tests/fixtures/variance.rb:24
     24 |  def inv(Invariant::[B] x) => nil; end
                               ^ B, with:
        @ __ROOT__/tests/fixtures/variance.rb:27
     27 |    inv(Invariant.new(A.new))
                                 ^^^ A

        - arising from an attempt to match:

        @ __ROOT__/tests/fixtures/variance.rb:24
     24 |  def inv(Invariant::[B] x) => nil; end
                   ^^^^^^^^^^^^^^ Invariant::[B], with:
        @ __ROOT__/tests/fixtures/variance.rb:27
     27 |    inv(Invariant.new(A.new))
                 ^^^^^^^^^^^^^^^^^^^^ Invariant::[A]
        @ __ROOT__/tests/fixtures/variance.rb:27
     27 |    inv(Invariant.new(A.new))
             ^^^^^^^^^^^^^^^^^^^^^^^^^ in this expression

error: Could not match types:

        @ __ROOT__/tests/fixtures/variance.rb:29
     29 |    inv(Invariant.new(C.new))
                                 ^^^ C, with:
        @ __ROOT__/tests/fixtures/variance.rb:24
     24 |  def inv(Invariant::[B] x) => nil; end
                               ^ B

        - arising from an attempt to match:

        @ __ROOT__/tests/fixtures/variance.rb:24
     24 |  def inv(Invariant::[B] x) => nil; end
                   ^^^^^^^^^^^^^^ Invariant::[B], with:
        @ __ROOT__/tests/fixtures/variance.rb:29
     29 |    inv(Invariant.new(C.new))
                 ^^^^^^^^^^^^^^^^^^^^ Invariant::[C]
        @ __ROOT__/tests/fixtures/variance.rb:29
     29 |    inv(Invariant.new(C.new))
             ^^^^^^^^^^^^^^^^^^^^^^^^^ in this expression

error: Could not match types:

        @ __ROOT__/tests/fixtures/variance.rb:32
     32 |  def cov(Covariant::[B] x) => nil; end
                               ^ B, with:
        @ __ROOT__/tests/fixtures/variance.rb:35
     35 |    cov(Covariant.new(A.new))
                                 ^^^ A

        - arising from an attempt to match:

        @ __ROOT__/tests/fixtures/variance.rb:32
     32 |  def cov(Covariant::[B] x) => nil; end
                   ^^^^^^^^^^^^^^ Covariant::[B], with:
        @ __ROOT__/tests/fixtures/variance.rb:35
     35 |    cov(Covariant.new(A.new))
                 ^^^^^^^^^^^^^^^^^^^^ Covariant::[A]
        @ __ROOT__/tests/fixtures/variance.rb:35
     35 |    cov(Covariant.new(A.new))
             ^^^^^^^^^^^^^^^^^^^^^^^^^ in this expression

error: Could not match types:

        @ __ROOT__/tests/fixtures/variance.rb:45
     45 |    contra(Contravariant.new(C.new))
                                        ^^^ C, with:
        @ __ROOT__/tests/fixtures/variance.rb:40
     40 |  def contra(Contravariant::[B] x) => nil; end
                                      ^ B

        - arising from an attempt to match:

        @ __ROOT__/tests/fixtures/variance.rb:40
     40 |  def contra(Contravariant::[B] x) => nil; end
                      ^^^^^^^^^^^^^^^^^^ Contravariant::[B], with:
        @ __ROOT__/tests/fixtures/variance.rb:45
     45 |    contra(Contravariant.new(C.new))
                    ^^^^^^^^^^^^^^^^^^^^^^^^ Contravariant::[C]
        @ __ROOT__/tests/fixtures/variance.rb:45
     45 |    contra(Contravariant.new(C.new))
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ in this expression
